File: shortest-path.cpp
#include <iostream>
#include <cstdio>
using namespace std;

class Edge
{
private:
	static Edge **head;
	static int n;
	Edge *_next = nullptr;
	int endPoint = -1, value = 0;

public:
	Edge(Edge *_next = nullptr, int endPoint = -1, int value = 0)
	{
		this->_next = _next;
		this->endPoint = endPoint;
		this->value = value;
	}
	int getEndPoint()
	{
		return this->endPoint;
	}
	int getValue()
	{
		return this->value;
	}
	Edge *nextEdge()
	{
		return this->_next;
	}
};

class Graph
{
private:
	Edge **head = nullptr;
	int n = 0, m = 0;

public:
	Graph(int n = 0, int m = 0)
	{
		this->n = n;
		this->m = m;
		head = new Edge *[n];
		for (int i = 0; i < n; i++)
			head[i] = nullptr;
	}
	void addEdge(int u, int v, int w)
	{
		Edge *tmp = new Edge(head[u], v, w);
		head[u] = tmp;
	}
	int numOfNodes()
	{
		return this->n;
	}
	int numOfEdges()
	{
		return this->m;
	}
	Edge *firstEdge(int u)
	{
		return head[u];
	}
	Edge *nextEdge(Edge *i)
	{
		return i->nextEdge();
	}
	void print()
	{
		for (int i = 0; i < n; i++)
		{
			cout << i << ": ";
			for (Edge *j = firstEdge(i); j != nullptr; j = nextEdge(j))
			{
				cout << j->getEndPoint() << ' ';
			}
			cout << endl;
		}
	}
};

typedef pair<int, int> PII;

class Heap
{
private:
	int sz = 0;
	PII *data = nullptr;
	void checkdown(int x)
	{
		if (x > sz)
			return;
		if ((x * 2 <= sz && data[x].second < data[x * 2].second) && (x * 2 + 1 <= sz && data[x].second < data[x * 2 + 1].second))
			return;
		if (x * 2 <= sz && data[x * 2].second < data[x * 2 + 1].second)
		{
			swap(data[x], data[x * 2]);
			checkdown(x * 2);
		}
		else if (x * 2 + 1 <= sz)
		{
			swap(data[x], data[x * 2 + 1]);
			checkdown(x * 2 + 1);
		}
	}

public:
	Heap(int sz = 0)
	{
		data = new PII[sz + 1];
	}
	void push(PII x)
	{
		data[++sz] = x;
		int t = sz;
		while (t != 1 && data[t].second < data[t / 2].second)
		{
			swap(data[t], data[t / 2]);
			t /= 2;
		}
	}
	PII top()
	{
		return data[1];
	}
	void pop()
	{
		data[1] = data[sz];
		sz--;
		checkdown(1);
	}
	bool empty()
	{
		if (sz < 1)
			return true;
		else
			return false;
	}
};

int dijkstra(Graph G, int s, int t)
{
	const int Inf = 0x3fffffff;
	int *dis = new int[G.numOfNodes()];
	bool *used = new bool[G.numOfNodes()];

	for (int i = 0; i < G.numOfNodes(); i++)
		dis[i] = Inf, used[i] = false;
	dis[s] = 0;
	Heap q(G.numOfNodes() * G.numOfEdges());
	q.push({s, 0});

	while (!q.empty())
	{
		PII top = q.top();
		q.pop();
		s = top.first;
		if (used[s])
			continue;
		used[s] = true;

		for (Edge *j = G.firstEdge(s); j != nullptr; j = j->nextEdge())
		{
			int v = j->getEndPoint(), w = j->getValue();
			if (!used[v] && dis[s] + w < dis[v])
			{
				dis[v] = dis[s] + w;
				q.push({v, dis[v]});
			}
		}
	}
	return dis[t];
}

int main()
{
	int n, m, u, v, w, s, t;
	scanf("%d%d%d%d", &n, &m, &s, &t);
	Graph G(n, m);

	while (m--)
	{
		scanf("%d%d%d", &u, &v, &w);
		G.addEdge(u, v, w);
		G.addEdge(v, u, w);
	}

	cout << dijkstra(G, s, t);
	return 0;
}

File: libs/shortest-path.js
/**
 * 边
 */
class Edge {
	endPoint = -1;
	value = 0;
	constructor(endPoint = -1, value = 0) {
		this.endPoint = endPoint;
		this.value = value;
	}
}

/**
 * 邻接表
 */
class Graph {
	head = null;
	nodes = 0;
	edges = 0;
	constructor(n, m) {
		this.nodes = n;
		this.edges = m;
		this.head = new Array(n);
	}

	/**
	 * 为邻接表添加新的边
	 * @param {Number} u - 起点
	 * @param {Number} v - 终点
	 * @param {Number} w - 边的权值
	 */
	addEdge(u, v, w) {
		if (this.head[u] === undefined) this.head[u] = new Array();
		this.head[u].push(new Edge(v, w));
	}
}

/**
 * 小根堆
 */
class Heap {
	#data = new Array();
	constructor() {
		this.#data.push(new Edge(-1, 0));
	}

	/**
	 * 将堆的顶点下沉到对应位置
	 * @param {Number} x - 堆的节点编号
	 */
	checkdown(x) {
		let size = this.#data.length - 1;
		let t1 = x * 2;
		let t2 = x * 2 + 1;
		if (x > size) return;
		if ((t1 <= size && this.#data[x].value < this.#data[t1].value) && (t2 <= size && this.#data[x].value < this.#data[t2].value)) return;
		if ((t2 <= size && this.#data[t1].value < this.#data[t2].value) || (t1 == size)) {
			let tmp = this.#data[x];
			this.#data[x] = this.#data[t1];
			this.#data[t1] = tmp;
			this.checkdown(x * 2);
		}
		else if (t2 <= size) {
			let tmp = this.#data[x];
			this.#data[x] = this.#data[t2];
			this.#data[t2] = tmp;
			this.checkdown(t2);
		}
	}

	/**
	 * 向堆中添加新的Edge类元素
	 * @param {Edge} obj 
	 */
	push(obj) {
		this.#data.push(obj);
		let t1 = this.#data.length - 1;
		let t2 = parseInt(t1 / 2);
		while (t1 != 1 && this.#data[t1].value < this.#data[t2].value) {
			let tmp = this.#data[t1];
			this.#data[t1] = this.#data[t2];
			this.#data[t2] = tmp;
			t1 = t2;
			t2 = parseInt(t1 / 2);
		}
	}

	/**
	 * 将堆顶元素移出堆
	 */
	pop() {
		let size = this.#data.length - 1;
		this.#data[1] = this.#data[size];
		this.#data.pop();
		this.checkdown(1);
	}

	/**
	 * 堆是否为空
	 */
	get empty() {
		let size = this.#data.length - 1;
		if (size < 1) return true;
		else return false;
	}

	/**
	 * 返回堆顶元素
	 */
	get top() {
		return this.#data[1];
	}
}

/**
 * 堆优化的dijkstra算法
 * @param {Graph} G - 无向图
 * @param {Number} s - 起点
 * @param {Number} t - 终点
 * @returns 最短路径长度以及路线
 */
function dijkstra(G, s, t) {
	let dis = new Array(G.nodes);
	let used = new Array(G.nodes);
	let pre = new Array(G.nodes);

	for (let i = 0; i < G.nodes; i++) {
		dis[i] = Infinity;
		used[i] = false;
		pre[i] = -1;
	}

	dis[s] = 0;
	let q = new Heap();
	q.push(new Edge(s, 0));

	while (!q.empty) {
		let top = q.top;
		q.pop();
		s = top.endPoint;
		if (used[s]) continue;
		used[s] = true;

		for (let i in G.head[s]) {
			let v = G.head[s][i].endPoint;
			let w = G.head[s][i].value;
			if (!used[v] && dis[s] + w < dis[v]) {
				pre[v] = s;
				dis[v] = dis[s] + w;
				q.push(new Edge(v, dis[v]));
			}
		}
	}

	return { pathway: pre, distance: dis };
}

/**
 * 调用dijkstra算法
 * @param {Object} obj - 无向图
 * @param {Number} s - 起点
 * @returns 最短路径长度以及路线
 */
function calcShortestPath(obj, s) {
	let G = new Graph(obj.nodes.length, obj.edges.length);
	for (let i in obj.edges) {
		if (obj.edges[i].alive) {
			G.addEdge(obj.edges[i].source, obj.edges[i].target, obj.edges[i].monster.damage);
			G.addEdge(obj.edges[i].target, obj.edges[i].source, obj.edges[i].monster.damage);
		}
		else {
			G.addEdge(obj.edges[i].source, obj.edges[i].target, 0);
			G.addEdge(obj.edges[i].target, obj.edges[i].source, 0);
		}
	}
	return dijkstra(G, s);
}
